alembic/ddl/impl.py

```diff
@@ -322,32 +328,63 @@ def bulk_insert(self, table, rows, multiinsert=True):
                     for row in rows:
                         self._exec(table.insert(inline=True).values(**row))
 
-    def compare_type(self, inspector_column, metadata_column):
-
-        conn_type = inspector_column.type
-        metadata_type = metadata_column.type
+    def _tokenize_column_type(self, column):
+        definition = self.dialect.type_compiler.process(column.type).lower()
+        # py27 - py36 - col_type, *param_terms = re.findall...
+        matches = re.findall("[^(),]+", definition)
+        col_type, param_terms = matches[0], matches[1:]
+        params = Params(col_type, [], {})
+        for term in param_terms:
+            if "=" in term:
+                key, val = term.split("=")
+                params.kwargs[key.strip()] = val.strip()
+            else:
+                params.args.append(term.strip())
+        return params
 
-        metadata_impl = metadata_type.dialect_impl(self.dialect)
-        if isinstance(metadata_impl, sqltypes.Variant):
-            metadata_impl = metadata_impl.impl.dialect_impl(self.dialect)
+    def _column_types_match(self, inspector_type, metadata_type):
+        if inspector_type == metadata_type:
+            return True
+        synonyms = [{t.lower() for t in batch} for batch in self.type_synonyms]
+        for batch in synonyms:
+            if {inspector_type, metadata_type}.issubset(batch):
+                return True
+        return False
 
-        # work around SQLAlchemy bug "stale value for type affinity"
-        # fixed in 0.7.4
-        metadata_impl.__dict__.pop("_type_affinity", None)
+    def _column_args_match(self, inspected_params, meta_params):
+        """We want to compare column parameters. However, we only want
+        to compare parameters that are set. If they both have `collation`,
+        we want to make sure they are the same. However, if only one
+        specifies it, dont flag it for being less specific
+        """
+        # py27- .keys is a set in py36
+        for kw in (
+            set(inspected_params.kwargs.keys())
+            & set(meta_params.kwargs.keys())
+        ):
+            if inspected_params.kwargs[kw] != meta_params.kwargs[kw]:
+                return False
 
-        if hasattr(metadata_impl, "compare_against_backend"):
-            comparison = metadata_impl.compare_against_backend(
-                self.dialect, conn_type
-            )
-            if comparison is not None:
-                return not comparison
+        for meta, inspect in zip(inspected_params.args, meta_params.args):
```

Comment:
**mike bayer** (zzzeek) wrote:

do we know that these two .args are the same length?  What if a VARCHAR has a default length that gets set up on the database side but isn't in the metadata?  i think perhaps if len(meta_params.args) == 0, we skip this test.

Reply:
The reason I picked zip was in case was in case the lists weren't of equal length. We only compare up to the length of the shorter list, ignoring differences in "extra" arguments (that may have been added by the server defaults). If the shorter list is len 0, this loop will just be skipped.

Reply:
do we have any scenarios that exercise these cases?  I';m not sure our datatypes have so much positional arguments for this to matter.  I'm fine with it for now.

Reply:
If I do

```
if inspected_params.args and meta_params.args and inspected_params.args != meta_params.args:
    return False
```

AutogenerateVariantCompareTest_oracle+cx_oracle_11_2_0_2_0::test_variant_no_issue fails comparing BigInteger to BigInteger, while test_autogen_diffs.py::CompareMetadataToInspectorTest_oracle+cx_oracle_11_2_0_2_0::test_introspected_columns_match_metadata_columns[cola10] fails comparing Boolean to Boolean. 
This is because the Oracle dialect prints those as NUMERIC(19) ... but the server adds precision to the returned value.

I added an extra test to be a bit more explicit with DECIMAL(10, 2) compared to NUMERIC(10)